File: Java/CalculadoraBásica/Calculadora.java
package CalculadoraBásica;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
/**
 * Calculadora es una clase que implementa una calculadora básica en modo consola.
 * Permite realizar operaciones de suma, resta, multiplicación, división y potenciación.
 * Utiliza el Objeto BufferedReader para leer la entrada del usuario y maneja excepciones para
 * evitar errores comunes como entradas no numéricas y división por cero.
 * @author MarioKiller44
 * @version 1.0.3
 * @since 2023-10-01
 */
public class Calculadora {

    /**
     * Este atributo se utiliza para controlar el flujo del programa y determinar qué operación realizar.
     * Atributo que almacena la opción seleccionada por el usuario.
     * Inicialmente se establece en -1 para indicar que no se ha seleccionado ninguna opción.
     * 
     */
    private static int opcion = -1;

    /**
     * Método principal que inicia la ejecución del programa.
     * Maneja excepciones para evitar errores comunes y proporciona un menú interactivo.
     * @version 1.0.1
     * @param args Argumentos de la línea de comandos 
     * @return void
     * @throws Exception Si ocurre un error al leer la entrada del usuario o al procesar
     * las operaciones, se captura y muestra un mensaje de error.
     */
    public static void main(String[] args) {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));

        System.out.println("Hola chicos, que tal. Este es mi primer programa de Java hecho en Visual Studio Code\n");
        boolean devo = false;
        while (devo != true) {
            /**
             * Bloque try-catch para manejar excepciones al leer la entrada del usuario.
             */ 
            try {

                System.out.println("Elige una opción de esta calculadora\n");
                System.out.println("1: Sumar\n2: Restar\n3: Multiplicar\n4: Dividir\n5: Potenciar\n6: Salir");
                System.out.print("Elija: ");
                String texto = bufferedReader.readLine();

                opcion = Integer.parseInt(texto);
                switch (opcion) {
                    case 1:
                        System.out.println("¡De acuerdo, vamos a sumar!");
                        suma();
                        break;
                    case 2:
                        System.out.println("¡De acuerdo, vamos a restar!");
                        resta();
                        break;
                    case 3:
                        System.out.println("¡De acuerdo, vamos a multiplicar!");
                        multiplicar();
                        break;
                    case 4:
                        System.out.println("¡De acuerdo, vamos a dividir!");
                        dividir();
                        break;
                    case 5:
                        System.out.println("¡De acuerdo, vamos a potenciar!");
                        potenciar();
                        break;
                    case 6:
                        System.out.println("¡De acuerdo, vamos a salir!");
                        System.exit(0);
                        break;

                    default:
                        System.out.println("Opción equivocada");
                        devo = false;
                        break;

                }
            } catch (Exception e) {
                System.out.println("Ha ocurrido un error, con el siguiente mensaje: " + e.getLocalizedMessage());

            }
        }
    }
//<editor-fold desc="Métodos de operaciones matemáticas" collapsed="true">
    private static boolean suma() {
        boolean devoSuma = false;
        ArrayList<Double> numeroList = new ArrayList<>();
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        boolean devoNum = false;
        Double numeroResultante = 0.0;
        String respuesta;
        try {
            while (!devoNum) {
                System.out.println("Introduce número a sumar:");
                numeroList.add(Double.parseDouble(bf.readLine()));
                if (numeroList.size() > 1) {
                    System.out.println("¿Seguimos sumando más? (Sí/No)");
                    respuesta = bf.readLine();
                    if (respuesta.equalsIgnoreCase("No")) {
                        devoNum = true;
                    }
                }
            }
            for (Double num : numeroList) {
                numeroResultante += num;
            }
            System.out.println("El resultado es: " + numeroResultante);
        } catch (Exception e) {
            System.out.println("Ha ocurrido un error en la suma, con el siguiente mensaje: " + e.getLocalizedMessage());
        }
        return devoSuma;
    }

    private static boolean resta() {
        boolean devoResta = false;
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        try {
            System.out.println("Introduce el primer número:");
            double num1 = Double.parseDouble(bf.readLine());
            System.out.println("Introduce el segundo número:");
            double num2 = Double.parseDouble(bf.readLine());
            double resultado = num1 - num2;
            System.out.println("El resultado es: " + resultado);
        } catch (Exception e) {
            System.out.println("Ha ocurrido un error en la resta, con el siguiente mensaje: " + e.getLocalizedMessage());
        }
        return devoResta;
    }

    private static boolean multiplicar() {
        boolean devoMultiplicar = false;
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        try {
            System.out.println("Introduce el primer número:");
            double num1 = Double.parseDouble(bf.readLine());
            System.out.println("Introduce el segundo número:");
            double num2 = Double.parseDouble(bf.readLine());
            double resultado = num1 * num2;
            System.out.println("El resultado es: " + resultado);
        } catch (Exception e) {
            System.out.println("Ha ocurrido un error en la multiplicación, con el siguiente mensaje: " + e.getLocalizedMessage());
        }
        return devoMultiplicar;
    }

    private static boolean dividir() {
        boolean devoDividir = false;
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        try {
            System.out.println("Introduce el dividendo:");
            double num1 = Double.parseDouble(bf.readLine());
            System.out.println("Introduce el divisor:");
            double num2 = Double.parseDouble(bf.readLine());
            if (num2 == 0) {
                System.out.println("No se puede dividir por cero.");
            } else {
                double resultado = num1 / num2;
                System.out.println("El resultado es: " + resultado);
            }
        } catch (Exception e) {
            System.out.println("Ha ocurrido un error en la división, con el siguiente mensaje: " + e.getLocalizedMessage());
        }
        return devoDividir;
    }

    private static boolean potenciar() {
        boolean devoPotenciar = false;
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        try {
            System.out.println("Introduce la base:");
            double base = Double.parseDouble(bf.readLine());
            System.out.println("Introduce el exponente:");
            double exponente = Double.parseDouble(bf.readLine());
            double resultado = Math.pow(base, exponente);
            System.out.println("El resultado es: " + resultado);
        } catch (Exception e) {
            System.out.println("Ha ocurrido un error en la potenciación, con el siguiente mensaje: " + e.getLocalizedMessage());
        }
        return devoPotenciar;
    }
//</editor-fold>

}

File: Pyhton/calculadora.py
import math;


def suma(a, b):
    return a + b
def resta(a, b):
    return a - b
def multiplicacion(a, b):
    return a * b
def division(a, b):
    if b != 0:
        return a / b
    else:
        return "Error: División por cero"
def potencia(a, b):
    return math.pow(a, b)
def raiz_cuadrada(a):
    if a >= 0:
        return math.sqrt(a)
    else:
        i=a
        return "Solucion en números complejos: "+str(complex(0,math.sqrt(-i)))
def raiz_cubica(a):
    return a ** (1/3)
def raiz_n(a, n):
    if n != 0:
        return a ** (1/n)
    else:
        return "Error: Raíz de índice cero"
def logaritmo(a, base=10):
    if a > 0 and base > 1:
        return math.log(a, base)
    else:
        return "Error: Logaritmo no definido para estos valores"
print("Bienvenido a la calculadora")
bandera=True;
while bandera:
    
    print("Seleccione la operación:")
    print("1. Suma")
    print("2. Resta")
    print("3. Multiplicación")
    print("4. División")
    print("5. Potencia")
    print("6. Raíz Cuadrada")
    print("7. Raíz Cúbica")
    print("8. Raíz enésima")
    print("9. Logaritmo\n")
    opcion = input()
    if opcion in ['1', '2', '3', '4', '5']:
        num1 = float(input("Ingrese el primer número: "))
        num2 = float(input("Ingrese el segundo número: "))
    if opcion == '1':
        print(f"{num1} + {num2} = {suma(num1, num2)}")
    elif opcion == '2':
        print(f"{num1} - {num2} = {resta(num1, num2)}")
    elif opcion == '3':
        print(f"{num1} * {num2} = {multiplicacion(num1, num2)}")
    elif opcion == '4':
        print(f"{num1} / {num2} = {division(num1, num2)}")
    elif opcion == '5':
        print(f"{num1} ^ {num2} = {potencia(num1, num2)}")
    elif opcion == '6':
        num = float(input("Ingrese el número: "))
        print(f"√{num} = {raiz_cuadrada(num)}")
    elif opcion == '7':
        num = float(input("Ingrese el número: "))
        print(f"∛{num} = {raiz_cubica(num)}")
    elif opcion == '8':
        num = float(input("Ingrese el número: "))
        n = float(input("Ingrese el índice de la raíz: "))
        print(f"{n}√{num} = {raiz_n(num, n)}")
    elif opcion == '9':
        num = float(input("Ingrese el número: "))
        base = float(input("Ingrese la base (default 10): ") or 10)
        print(f"log base {base} de {num} = {logaritmo(num, base)}")
    else:
        print("Opción inválida")
    respuesta = input("¿Desea realizar otra operación? (s/n): ")
    if respuesta.lower() != 's':
        bandera = False
        print("Gracias por usar la calculadora")
    else:
        print("\n")

    